scripts/affiliate-server

```diff
@@ -0,0 +1,222 @@
+#! /usr/bin/env python
+"""Run affiliate server.
+
+Usage:
+
+start affiliate-server using dev webserver:
+
+    ./scripts/affiliate-server openlibrary.yml 31337
+
+start affiliate-server as fastcgi:
+
+    ./scripts/affiliate-server openlibrary.yml fastcgi 31337
+
+start affiliate-server using gunicorn webserver:
+
+    ./scripts/affiliate-server openlibrary.yml --gunicorn -b 0.0.0.0:31337
+
+"""
+
+import json
+import logging
+import os
+import sys
+import time
+import web
+import yaml
+
+from openlibrary.core import cache
+from openlibrary.core.vendors import AmazonAPI
+from openlibrary.utils.isbn import (
+    normalize_isbn, isbn_13_to_isbn_10, isbn_10_to_isbn_13)
+from openlibrary.utils.dateutil import DAY_SECS
+import _init_path
+
+logger = logging.getLogger("affiliate-server")
+
+urls = (
+    '/isbn/([0-9X-]+)', 'Submit'
+)
+
+class Submit:
+
+    @classmethod
+    def unpack_isbn(cls, isbn):
+        isbn = normalize_isbn(isbn.replace('-', ''))
+        isbn10 = isbn if len(isbn) == 10 else isbn.startswith('978') and isbn_13_to_isbn_10(isbn)
+        isbn13 = isbn if len(isbn) == 13 else isbn_10_to_isbn_13(isbn)
+        return isbn10, isbn13
+
+
+    def GET(self, isbn, threshold=4, wait=3):
+        """
+        If the threshold is met, piggy back request and do an amazon lookup to fill cache.
+        Next, check if solution is in cache. If so, return it.
+        Otherwise, return the equivalent of a promise e.g. "submitted"
+
+        Wait until queue exceeds threshold length or until wait period
+        exceededed (i.e. wait seconds passed)
+        """
+
+        if not web.amazon_api:
+            return json.dumps({"error": "not_configured"})
+
+        isbn10, isbn13 = self.unpack_isbn(isbn)
+        if not (isbn10 and isbn13):
+            return json.dumps({
+                "error": "rejected_isbn",
+                "isbn10": isbn10,
+                "isbn13": isbn13
+            })
+
+        # Cache lookup by isbn13. If there's a miss, add to queue
+        if not cache.memcache_cache.get('amazon_product_%s' % isbn13):
+            # Misses (which will get submitted to amz as ASIN) are isbn10
+            # We should track when (queue % 20) occurs
```

Comment:
20?

Reply:
Sorry -- this was an arbitrary attempt at achieving the same thing as sampling statsd (e.g. every time the queue passes some multiple of 20)